File: AirshipKit/AirshipKit/UAMediaEventTemplate.m
/*
 Copyright 2009-2016 Urban Airship Inc. All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice, this
 list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.

 THIS SOFTWARE IS PROVIDED BY THE URBAN AIRSHIP INC ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 EVENT SHALL URBAN AIRSHIP INC OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "UAMediaEventTemplate.h"
#import "UAirship.h"
#import "UAAnalytics.h"
#import "UACustomEvent+Internal.h"

#define kUAMediaEventTemplate @"media"
#define kUABrowsedContentEvent @"browsed_content"
#define kUAConsumedContentEvent @"consumed_content"
#define kUAStarredContentEvent @"starred_content"
#define kUASharedContentEvent @"shared_content"
#define kUAMediaEventTemplateLifetimeValue @"ltv"
#define kUAMediaEventTemplateIdentifier @"id"
#define kUAMediaEventTemplateCategory @"category"
#define kUAMediaEventTemplateDescription @"description"
#define kUAMediaEventTemplateType @"type"
#define kUAMediaEventTemplateFeature @"feature"
#define kUAMediaEventTemplateAuthor @"author"
#define kUAMediaEventTemplatePublishedDate @"published_date"
#define kUAMediaEventTemplateSource @"source"
#define kUAMediaEventTemplateMedium @"medium"

@interface UAMediaEventTemplate()
@property (nonatomic, copy) NSString *eventName;
@property (nonatomic, copy) NSString *source;
@property (nonatomic, copy) NSString *medium;
@property (nonatomic, strong) NSDecimalNumber *eventValue;
@property (nonatomic, assign) BOOL featureSet;
@end

@implementation UAMediaEventTemplate

- (instancetype)initWithName:(NSString *)name
                   withValue:(NSDecimalNumber *)eventValue
                  withSource:(NSString *)source
                  withMedium:(NSString *)medium {
    self = [super init];
    if (self) {
        self.eventName = name;
        self.eventValue = eventValue;
        self.source = source;
        self.medium = medium;
    }

    return self;
}

+ (instancetype)browsedTemplate {
    return [[self alloc] initWithName:kUABrowsedContentEvent
                            withValue:nil
                           withSource:nil
                           withMedium:nil];
}

+ (instancetype)starredTemplate {
    return [[self alloc] initWithName:kUAStarredContentEvent
                            withValue:nil
                           withSource:nil
                           withMedium:nil];
}

+ (instancetype)sharedTemplateWithSource:(NSString *)source
                                  withMedium:(NSString *)medium {
    return [[self alloc] initWithName:kUASharedContentEvent
                            withValue:nil
                           withSource:source
                           withMedium:medium];
}

+ (instancetype)sharedTemplate {
    return [self sharedTemplateWithSource:nil withMedium:nil];
}

+ (instancetype)consumedTemplate {
    return [self consumedTemplateWithValue:nil];
}

+ (instancetype)consumedTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self consumedTemplateWithValue:decimalValue];
}

+ (instancetype)consumedTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUAConsumedContentEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil];
}

- (void)setIsFeature:(BOOL)isFeature {
    self.featureSet = YES;
    _isFeature = isFeature;
}

- (UACustomEvent *)createEvent {
    UACustomEvent *event = [UACustomEvent eventWithName:self.eventName];

    if (self.eventValue) {
        [event setEventValue:self.eventValue];
        [event setBoolProperty:YES forKey:kUAMediaEventTemplateLifetimeValue];
    } else {
        [event setBoolProperty:NO forKey:kUAMediaEventTemplateLifetimeValue];
    }

    if (self.identifier) {
        [event setStringProperty:self.identifier forKey:kUAMediaEventTemplateIdentifier];
    }

    if (self.category) {
        [event setStringProperty:self.category forKey:kUAMediaEventTemplateCategory];
    }

    if (self.eventDescription) {
        [event setStringProperty:self.eventDescription forKey:kUAMediaEventTemplateDescription];
    }

    if (self.type) {
        [event setStringProperty:self.type forKey:kUAMediaEventTemplateType];
    }

    if (self.featureSet) {
        [event setBoolProperty:self.isFeature forKey:kUAMediaEventTemplateFeature];
    }

    if (self.author) {
        [event setStringProperty:self.author forKey:kUAMediaEventTemplateAuthor];
    }

    if (self.publishedDate) {
        [event setStringProperty:self.publishedDate forKey:kUAMediaEventTemplatePublishedDate];
    }

    if (self.source) {
        [event setStringProperty:self.source forKey:kUAMediaEventTemplateSource];
    }

    if (self.medium) {
        [event setStringProperty:self.medium forKey:kUAMediaEventTemplateMedium];
    }

    event.templateType = kUAMediaEventTemplate;
    return event;
}

@end

File: Airship/AirshipCore/Source/UAMediaEventTemplate.swift
/* Copyright Airship and Contributors */

/**
 * A UAMediaEventTemplate represents a custom media event template for the
 * application.
 */
@objc
public class UAMediaEventTemplate : NSObject {
    /**
     * The event's ID.
     */
    @objc
    public var identifier : String?

    /**
     * The event's category.
     */
    @objc
    public var category : String?

    /**
     * The event's type.
     */
    @objc
    public var type : String?

    /**
     * The event's description.
     */
    @objc
    public var eventDescription : String?

    private var _isFeature: Bool?

    /**
     * `YES` if the event is a feature, else `NO`.
     */
    @objc
    public var isFeature : Bool {
        get {
            return self._isFeature ?? false
        }
        set {
            self._isFeature = newValue
        }
    }

    /**
     * The event's author. The author's length must not exceed 255 characters
     * or it will invalidate the event.
     */
    @objc
    public var author : String?

    /**
     * The event's publishedDate. The publishedDate's length must not exceed 255 characters
     * or it will invalidate the event.
     */
    @objc
    public var publishedDate : String?

    private let eventName: String
    private let medium: String?
    private let source: String?
    private let eventValue: NSNumber?

    /**
     * Factory method for creating a browsed content event template.
     * @returns A Media event template instance
     */
    @objc
    public class func browsedTemplate() -> UAMediaEventTemplate {
        return UAMediaEventTemplate("browsed_content")
    }


    /**
     * Factory method for creating a starred content event template.
     * @returns A Media event template instance
     */
    @objc
    public class func starredTemplate() -> UAMediaEventTemplate {
        return UAMediaEventTemplate("starred_content")

    }


    /**
     * Factory method for creating a shared content event template.
     * @returns A Media event template instance
     */
    @objc
    public class func sharedTemplate() -> UAMediaEventTemplate {
        return sharedTemplate(source: nil, medium: nil)
    }


    /**
     * Factory method for creating a shared content event template.
     * If the source or medium exceeds 255 characters it will cause the event to be invalid.
     *
     * @param source The source as an NSString.
     * @param medium The medium as an NSString.
     * @returns A Media event template instance
     */
    @objc(sharedTemplateWithSource:withMedium:)
    public class func sharedTemplate(source: String?, medium: String?) -> UAMediaEventTemplate {
        return UAMediaEventTemplate("shared_content", value: nil, source: source, medium: medium)
    }


    /**
     * Factory method for creating a consumed content event template.
     * @returns A Media event template instance
     */
    @objc
    public class func consumedTemplate() -> UAMediaEventTemplate {
        return consumedTemplate(value: nil)
    }


    /**
     * Factory method for creating a consumed content event template with a value.
     *
     * @param valueString The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Media event template instance
     */
    @objc(consumedTemplateWithValueFromString:)
    public class func consumedTemplate(valueString: String?) -> UAMediaEventTemplate {
        let decimalValue = valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return consumedTemplate(value: decimalValue)
    }


    /**
     * Factory method for creating a consumed content event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @returns A Media event template instance
     */
    @objc(consumedTemplateWithValue:)
    public class func consumedTemplate(value: NSNumber?) -> UAMediaEventTemplate {
        return UAMediaEventTemplate("consumed_content", value: value)
    }

    private init(_ eventName: String, value: NSNumber? = nil, source: String? = nil, medium: String? = nil) {
        self.eventName = eventName
        self.eventValue = value
        self.source = source
        self.medium = medium
        super.init()
    }

    /**
     * Creates the custom media event.
     */
    @objc
    public func createEvent() -> UACustomEvent {
        var propertyDictionary: [String : Any] = [:]
        propertyDictionary["ltv"] = self.eventValue != nil
        propertyDictionary["id"] = self.identifier
        propertyDictionary["category"] = self.category
        propertyDictionary["type"] = self.type
        propertyDictionary["feature"] = self.isFeature
        propertyDictionary["published_date"] = self.publishedDate
        propertyDictionary["source"] = self.source
        propertyDictionary["medium"] = self.medium
        propertyDictionary["description"] = self.eventDescription
        propertyDictionary["author"] = self.author

        let event = UACustomEvent(name: self.eventName)
        event.templateType = "media"
        event.eventValue = self.eventValue
        event.properties = propertyDictionary
        return event
    }
}
